firedrake/solving_utils.py

```diff
@@ -70,10 +71,13 @@ def __init__(self, problem, mat_type, pmat_type, appctx=None,
                  pre_jacobian_callback=None, pre_function_callback=None,
                  options_prefix=None):
         from firedrake.assemble import allocate_matrix, create_assembly_callable
+        self.allocate_matrix = allocate_matrix
+        self.create_assembly_callable = create_assembly_callable
```

Comment:
These don't need to by instance attributes, you can instead import the relevant function in the appropriate `@cached_property`.